### ISSUE TITLE
[Feature] Add a hook before yarn add

### ISSUE THREAD

**Issue by staldera** [yarnpkg/berry#6152]
- [ ] I'd be willing to implement this feature ([contributing guide](https://yarnpkg.com/advanced/contributing))
- [x] This feature is important to have in this repository; a contrib plugin wouldn't do

**Describe the user story**

As a developer I want to run a vulnerability check and some extra checks BEFORE adding a package with `yarn add` (eg: run a tool such as https://github.com/lirantal/npq).
Thus, I looked into hooks, but I could only find one that's called before add, and it's `validateProject`. Sadly this one is called very often and doesn't contain the original command in the parameters (so no way to know if it's called as part of `yarn install` or `yarn add` or `yarn <something else>`).

**Describe the solution you'd like**

I'd like to have a hook that is called before adding a package, and has a parameter that gives me the packages that are supposed to be added. This way I can run vulnerability checks (eg: `yarn npq add <package1> <package2> --dry-run`) and proceed (if no problems) or signal a problem/exit/throw.

Having a `report` parameter such as there is in multiple existing hooks would be nice, because that'd allow me to use it to stop the current command cleanly if a vulnerability was found.

Ideally, I also would like to add a parameter such as `--force` to proceed despite warnings, but that might be an issue for `yarn add` command itself. Willing to do some extra magic myself to handle this case (eg: using `process.env` vars or a not-so-nice syntax such as `yarn add force <package1> <package2>`)

**Describe the drawbacks of your solution**

User-wise, I think there's no drawbacks, as you wouldn't need to use the hook if you don't need it.

**Describe alternatives you've considered**

I considered other alternatives, such as
- Replacing the add command completely by a custom one: Not possible AFAIK, not with scripts in `package.json`, nor with a command based on `BaseCommand`. If it was possible, that'd also fit my use-case though.
- Adding a custom command (eg: `addsafe`) and blocking the `add` command (through an error message): AFAIK also not possible to block a command currently. Would be less convenient than the first option, and might not work, since in the end (if the error vulnerability report is ok) I still want to `add` the package.

**Comment by jonsherrard** [comment]
I would also like this

I'd like to check that `yarn add` `yarn` or `yarn install` are being run with the the package.json `packageManager` version that is specificied, so I don't run into immutable errors on CI when the yarn.lock ends up changing.

**Comment by staldera** [comment]
Hi, thanks a lot for the feedback! I managed to do most of what I wanted to do with those two hooks and storing the results outside the hook to report them in validateProject, but the idea to start a report inside the hook sounds a bit cleaner to me, how would I do that?

**Comment by clemyan** [comment]
You can just `new` one of the Report classes (e.g. `StreamReport`), or use something like `StreamReport.start` which passes a report to a callback. See [`installWithNewReport`](https://github.com/yarnpkg/berry/blob/e06bacdb8091b7a25fdb7911c3466184b94fa040/packages/yarnpkg-core/sources/Project.ts#L1737) for an example.

**Comment by staldera** [comment]
I managed this:

```
          async afterWorkspaceDependencyAddition(workspace, target, descriptor) {
            const report = await StreamReport.start({
              configuration: workspace.project.configuration,
              json: false,
              stdout: Cli.defaultContext.stdout,
              forceSectionAlignment: true,
              includeLogs: true,
              includeVersion: true,
            }, async report => {
              const npqIssues = checkDependency(descriptor.scope, descriptor.name, descriptor.range, report)
              if (npqIssues) {
                if(process.env.NPQ_FORCE_INSTALL) {
                  report.reportWarning(MessageName.DANGEROUS_NODE_MODULES, `Issues were found with some package(s), but force mode is set. Proceeding:\n${npqIssues}`)
                }
                else {
                  report.reportError(MessageName.DANGEROUS_NODE_MODULES, `Issues were found with some package(s), use yarn add-force [package(s)] or yarn up-force [package(s)] if you still want to proceed:\n${npqIssues}`)
                }
              }
            });
          },
```

However, unlike when I use reportError in validateProject, the process doesn't exit. Is there something I missed or is that intentional?

**Comment by staldera** [comment]
Closed the issue by accident, sorry :(

**Comment by clemyan** [comment]
Yes. The "report" you get in install hooks (e.g. `validateProject`) is a special one that aborts the install if `reportError` is called. In `afterWorkspaceDependencyAddition`, you need to actually throw to abort. For `StreamReport` you can check `report.hasErrors()` to see if `reportError` has been called.
